## Machine learning Model Development

- In this section, we will develop and evaluate machine learning models to predict life expectancy based on various public health metrics, environmental factors, and demographic data. The process will be carried out in a systematic manner, starting from data preparation, model training, and evaluation, followed by model tuning and final evaluation.




In [4]:
# importing libraries and prepared dataset

import pandas as pd
from sklearn.model_selection import train_test_split

data = "/Users/alexandreribeiro/Documents/GitHub/final_project/notebooks/dataset_for_ml.csv"

df = pd.read_csv(data)

df.sample(5)

population_city  greenspacearea_km2       AQI  adjusted_obesity_rate  \
2626        -0.717821           -0.334645  0.131252               1.601020   
2333        -0.623986            0.002107  1.196798              -1.412754   
263          1.358519           -0.401819  0.014227               1.496306   
2560        -0.704587            1.308919  0.106615               0.162706   
2759        -0.753298            1.055639  0.556239              -0.208040   

      adjusted_smoking_rate  adjusted_copd_rate  adjusted_depression_rate  \
2626               1.577733           -0.100460                 -0.333365   
2333              -1.353649           -1.330394                 -1.342325   
263                1.841403            1.857407                  1.520905   
2560               0.124943            0.039512                  0.903436   
2759              -0.557730           -0.568735                 -0.394449   

      adjusted_life_expectancy  
2626                  1.321772  
2333                 -1.392546  
263                   1.090962  
2560                 -0.017670  
2759                 -0.198041

#### Defining features and target variable

- The first step in developing a machine learning model is to define the features and target variable. In this case, the target variable is the life expectancy, and the features are the various public health metrics, environmental factors, and demographic data. We will use the following features to predict life expectancy.

In [5]:
# Selected features

selected_features = ['population_city', 'greenspacearea_km2', 'AQI', 'adjusted_obesity_rate',
       'adjusted_smoking_rate', 'adjusted_copd_rate',
       'adjusted_depression_rate']

# Define the target variable (y) and the feature set (X)

y = df['adjusted_life_expectancy']  # Target variable
X = df[selected_features]  # Using the selected features

# Display the shapes of X and y

X.shape, y.shape

((3238, 7), (3238,))

#### Spliting the data into training and testing sets

- Split the dataset into training and testing sets to train the model on one portion and test it on another.

- test_size=0.2: This means that 20% of the data will be used for testing, and the remaining 80% will be used for training.
- random_state=42: Setting a random state ensures that the split is reproducible.

In [6]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the shapes of the training and testing sets
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (2590, 7)
X_test shape: (648, 7)
y_train shape: (2590,)
y_test shape: (648,)


#### Model Selection:

- We will start with a few common regression models since you’re predicting a continuous target variable (adjusted life expectancy).

We’ll try these models:

- Linear Regression
- Decision Tree Regressor
- Random Forest Regressor
- Gradient Boosting Regressor
- Support Vector Regressor (SVR)

- After selecting the models, we’ll compare their performance to choose the best one.

In [7]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Define the models
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "SVR": SVR()
}

# Dictionary to store the results
results = {}

# Train and evaluate each model
for model_name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the model
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    # Store the results
    results[model_name] = {
        "MAE": mae,
        "MSE": mse,
        "R²": r2
    }

    print(f"{model_name} results:")
    print(f"  - MAE: {mae:.3f}")
    print(f"  - MSE: {mse:.3f}")
    print(f"  - R²: {r2:.3f}\n")

# Convert the results to a DataFrame for easier comparison
import pandas as pd
results_df = pd.DataFrame(results).T
print(results_df)

Linear Regression results:
  - MAE: 0.149
  - MSE: 0.048
  - R²: 0.957

Decision Tree results:
  - MAE: 0.063
  - MSE: 0.025
  - R²: 0.977

Random Forest results:
  - MAE: 0.064
  - MSE: 0.016
  - R²: 0.985

Gradient Boosting results:
  - MAE: 0.121
  - MSE: 0.031
  - R²: 0.972

SVR results:
  - MAE: 0.125
  - MSE: 0.039
  - R²: 0.965

                        MAE       MSE        R²
Linear Regression  0.148989  0.047983  0.956544
Decision Tree      0.062692  0.025441  0.976959
Random Forest      0.064405  0.016251  0.985282
Gradient Boosting  0.120992  0.030632  0.972258
SVR                0.125336  0.039029  0.964653


In [8]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import pandas as pd

# Define the models
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "SVR": SVR()
}

# Dictionary to store the results
results = []

# Train and evaluate each model
for model_name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the model
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    # Store the results in the list
    results.append({
        "Model": model_name,
        "MAE": mae,
        "MSE": mse,
        "R²": r2
    })

# Convert the results to a DataFrame for easier comparison
results_df = pd.DataFrame(results)
results_df

Model       MAE       MSE        R²
0  Linear Regression  0.148989  0.047983  0.956544
1      Decision Tree  0.065927  0.027341  0.975239
2      Random Forest  0.064782  0.016589  0.984976
3  Gradient Boosting  0.120892  0.030541  0.972340
4                SVR  0.125336  0.039029  0.964653

#### Cross-validation:

- We will use cross-validation to evaluate the performance of the models. Cross-validation is a technique that splits the data into multiple subsets (folds) and trains the model on different combinations of these subsets. This helps to get a more accurate estimate of the model’s performance.

In [9]:
from sklearn.model_selection import cross_val_score
import numpy as np

# Define the model
model = RandomForestRegressor()  # You can choose the model you prefer

# Perform 5-fold cross-validation
cv_scores = cross_val_score(model, X_train, y_train, cv=5, scoring='r2')

# Output the cross-validation results
print("Cross-Validation R² Scores: ", cv_scores)
print("Mean R²: ", np.mean(cv_scores))
print("Standard Deviation of R²: ", np.std(cv_scores))

Cross-Validation R² Scores:  [0.98424967 0.98022191 0.98490611 0.98493518 0.98767759]
Mean R²:  0.984398091280141
Standard Deviation of R²:  0.002398508069151893


#### Results:

- Cross-Validation R² Scores: The individual R² scores across the 5 folds are all very high, indicating that the model is consistently performing well on different subsets of the data.

- Mean R²: The average R² score across the 5 folds is approximately 0.984, which suggests that the model explains about 98.4% of the variance in the target variable.

- Standard Deviation of R²: The standard deviation is very low (0.0021), indicating that the model’s performance is consistent across different data splits. This is a good sign as it shows that the model is not overly dependent on any particular subset of the data.


In [10]:
# Lets check the remaining models

# List of models to evaluate

models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "SVR": SVR()
}

# Dictionary to store the cross-validation results
cv_results = {}

# Perform cross-validation for each model
for model_name, model in models.items():
    # Perform cross-validation with 5 folds
    cv_scores = cross_val_score(model, X_train, y_train, cv=5, scoring='r2')
    
    # Store the mean and standard deviation of R² scores
    cv_results[model_name] = {
        "Mean R²": cv_scores.mean(),
        "Std Dev R²": cv_scores.std()
    }

# Convert the results to a DataFrame for easier comparison
cv_results_df = pd.DataFrame(cv_results).T
cv_results_df

Mean R²  Std Dev R²
Linear Regression  0.953557    0.004549
Decision Tree      0.971877    0.002512
Random Forest      0.984254    0.002288
Gradient Boosting  0.968374    0.001993
SVR                0.960934    0.004133

#### Using Hyperparameter Tuning to try to improve the model performance

- Using GridSearchCV to find the best hyperparameters for the model. GridSearchCV is a technique that searches for the best combination of hyperparameters by evaluating the model’s performance on different combinations of hyperparameters.

In [11]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

# Define the parameter grid for Random Forest
param_grid_rf = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'max_features': ['auto', 'sqrt', 'log2'],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Initialize the Random Forest Regressor
rf = RandomForestRegressor()

# Initialize GridSearchCV with 5-fold cross-validation
grid_search_rf = GridSearchCV(estimator=rf, param_grid=param_grid_rf, 
                              cv=5, scoring='r2', n_jobs=-1, verbose=2)

# Fit the GridSearchCV on the training data
grid_search_rf.fit(X_train, y_train)

# Best parameters and the corresponding score
best_params_rf = grid_search_rf.best_params_
best_score_rf = grid_search_rf.best_score_

print(f"Best Parameters for Random Forest: {best_params_rf}")
print(f"Best R² Score for Random Forest: {best_score_rf:.4f}")

Fitting 5 folds for each of 324 candidates, totalling 1620 fits


[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.0s
[CV] END max_depth=None, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=300; total time=   0.0s
[CV] END max_depth=None,

/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   1.0s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=2, n_estimators=300; total time=   1.2s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.7s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=5, n_estimators=200; total time=   1.2s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=5, n_estimators=300; total time=   1.5s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=2, n_estimators=200; total time=   1.1s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.6s
[CV] END max_depth=None, max_features=sqrt, min_samples_leaf=2, min_samples_split=10, n_estimators=300; total time=   1.6s
[CV] END max_depth=No

/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:540: FitFailedWarning: 
540 fits failed out of a total of 1620.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
462 fits failed with the following error:
Traceback (most recent call last):
  File "/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/model_selection/_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/base.py", line 1466, in wrapper
    estimator._validate_params()
  File "/opt/anaconda3/envs/env_test/lib/python3.11/site-packages/sklearn/base.py", line 666, in _validate_params
    validate_parameter_

Best Parameters for Random Forest: {'max_depth': 20, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Best R² Score for Random Forest: 0.9763


In [12]:
# Get the best model from GridSearchCV
best_rf = grid_search_rf.best_estimator_

# Make predictions on the test set
y_pred_rf = best_rf.predict(X_test)

# Evaluate the performance
mae_rf = mean_absolute_error(y_test, y_pred_rf)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print(f"Random Forest - Tuned Model Performance:")
print(f"  - MAE: {mae_rf:.3f}")
print(f"  - MSE: {mse_rf:.3f}")
print(f"  - R²: {r2_rf:.3f}")

Random Forest - Tuned Model Performance:
  - MAE: 0.091
  - MSE: 0.024
  - R²: 0.978


#### Overfitting check

- Compare the performance of the model on the training data versus the test data. A common method is to look at the difference between training and validation scores, especially during cross-validation.

In [13]:
# Assuming you've already done hyperparameter tuning and have the best model
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Train the best model with the entire training set
best_rf_model = RandomForestRegressor(
    max_depth=20, 
    max_features='log2', 
    min_samples_leaf=1, 
    min_samples_split=2, 
    n_estimators=300,
    random_state=42
)

best_rf_model.fit(X_train, y_train)

# Evaluate the model on the training data
y_train_pred = best_rf_model.predict(X_train)

# Calculate metrics on training data
train_mae = mean_absolute_error(y_train, y_train_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
train_r2 = r2_score(y_train, y_train_pred)

print("Random Forest - Training Data Performance:")
print(f"  - MAE: {train_mae:.3f}")
print(f"  - MSE: {train_mse:.3f}")
print(f"  - R²: {train_r2:.3f}\n")

Random Forest - Training Data Performance:
  - MAE: 0.033
  - MSE: 0.003
  - R²: 0.997



In [14]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Define the best model with the optimal hyperparameters
best_rf_model = RandomForestRegressor(
    max_depth=20, 
    max_features='log2', 
    min_samples_leaf=1, 
    min_samples_split=2, 
    n_estimators=300,
    random_state=42
)

# Train the model on the entire training set
best_rf_model.fit(X_train, y_train)

# Evaluate the model on the training data
y_train_pred = best_rf_model.predict(X_train)

# Calculate metrics on training data
train_mae = mean_absolute_error(y_train, y_train_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
train_r2 = r2_score(y_train, y_train_pred)

print("Random Forest - Training Data Performance:")
print(f"  - MAE: {train_mae:.3f}")
print(f"  - MSE: {train_mse:.3f}")
print(f"  - R²: {train_r2:.3f}\n")

# Evaluate the model on the test data
y_test_pred = best_rf_model.predict(X_test)

# Calculate metrics on test data
test_mae = mean_absolute_error(y_test, y_test_pred)
test_mse = mean_squared_error(y_test, y_test_pred)
test_r2 = r2_score(y_test, y_test_pred)

print("Random Forest - Test Data Performance:")
print(f"  - MAE: {test_mae:.3f}")
print(f"  - MSE: {test_mse:.3f}")
print(f"  - R²: {test_r2:.3f}\n")

Random Forest - Training Data Performance:
  - MAE: 0.033
  - MSE: 0.003
  - R²: 0.997

Random Forest - Test Data Performance:
  - MAE: 0.093
  - MSE: 0.025
  - R²: 0.977



#### Conclusion:

- A difference of 0.02 in R² is not a strong indication of overfitting. It’s more likely a sign that the model is performing well. Overfitting concerns arise more when there are larger discrepancies between training and test performance. In this case the model seems to be well-tuned, and the small difference are likely acceptable. 

#### Trying XGBoost Regressor

- XGBoost is a powerful and efficient implementation of the gradient boosting algorithm. It is known for its speed and performance, making it a popular choice for machine learning tasks. We will use the XGBoost Regressor to predict life expectancy and evaluate its performance.

In [15]:
import xgboost as xgb
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import GridSearchCV

# Initialize XGBoost model

xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)

# Train the model

xgb_model.fit(X_train, y_train)

# Make predictions on the test set

y_pred = xgb_model.predict(X_test)

# Evaluate the model

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Display the evaluation metrics in a dataframe

results = pd.DataFrame(data=[mae, mse, r2], index=['MAE', 'MSE', 'R²'], columns=['XGBoost'])

results



XGBoost
MAE  0.066697
MSE  0.014045
R²   0.987280

#### Hyperparameter Tuning for XGBoost Regressor

In [16]:
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.3],
    'subsample': [0.7, 0.8, 0.9],
    'colsample_bytree': [0.7, 0.8, 0.9]
}

grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid, cv=5, scoring='r2', n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

print("Best Parameters for XGBoost:", grid_search.best_params_)
print("Best R² Score for XGBoost:", grid_search.best_score_)

# Evaluate the best model on the test set
best_xgb_model = grid_search.best_estimator_
y_pred_best = best_xgb_model.predict(X_test)

# Calculate metrics on test data
best_mae = mean_absolute_error(y_test, y_pred_best)
best_mse = mean_squared_error(y_test, y_pred_best)
best_r2 = r2_score(y_test, y_pred_best)

# Display the evaluation metrics in a dataframe

results['XGBoost Tuned'] = [best_mae, best_mse, best_r2]

results

Fitting 5 folds for each of 243 candidates, totalling 1215 fits
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.1s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.0s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.1s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.1s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.7; total time=   0.1s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.8; total time=   0.1s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.8; total time=   0.1s
[CV] END colsample_bytree=0.7, learning_rate=0.01, max_depth=3, n_estimators=100, subsample=0.9; total time=   0.0s
[CV] END

XGBoost  XGBoost Tuned
MAE  0.066697       0.059276
MSE  0.014045       0.012146
R²   0.987280       0.989000

##### Results:

- MAE (Mean Absolute Error) decreased from 0.0667 to 0.0593, indicating that the average error between the predicted and actual values has reduced.

- MSE (Mean Squared Error) also decreased from 0.0140 to 0.0121, which means that the model’s overall error has been reduced, especially penalizing larger errors.

- R² (Coefficient of Determination) increased slightly from 0.9873 to 0.9890, showing that the model now explains slightly more variance in the data.

#### Checking for overfitting on XGBoost Regressor

In [17]:
import xgboost as xgb

# Best parameters from your hyperparameter tuning
best_params_xgb = {
    'colsample_bytree': 0.9,
    'learning_rate': 0.1,
    'max_depth': 7,
    'n_estimators': 300,
    'subsample': 0.9
}

# Initialize and train the XGBoost model with the best parameters
tuned_xgb_model = xgb.XGBRegressor(**best_params_xgb)
tuned_xgb_model.fit(X_train, y_train)

# Now, evaluate the tuned XGBoost model on both the training and test data
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Evaluate on the training data
y_train_pred_xgb = tuned_xgb_model.predict(X_train)
train_mae_xgb = mean_absolute_error(y_train, y_train_pred_xgb)
train_mse_xgb = mean_squared_error(y_train, y_train_pred_xgb)
train_r2_xgb = r2_score(y_train, y_train_pred_xgb)

print("Tuned XGBoost - Training Data Performance:")
print(f"  - MAE: {train_mae_xgb:.3f}")
print(f"  - MSE: {train_mse_xgb:.3f}")
print(f"  - R²: {train_r2_xgb:.3f}\n")

# Evaluate on the test data
y_test_pred_xgb = tuned_xgb_model.predict(X_test)
test_mae_xgb = mean_absolute_error(y_test, y_test_pred_xgb)
test_mse_xgb = mean_squared_error(y_test, y_test_pred_xgb)
test_r2_xgb = r2_score(y_test, y_test_pred_xgb)

print("Tuned XGBoost - Test Data Performance:")
print(f"  - MAE: {test_mae_xgb:.3f}")
print(f"  - MSE: {test_mse_xgb:.3f}")
print(f"  - R²: {test_r2_xgb:.3f}")

Tuned XGBoost - Training Data Performance:
  - MAE: 0.005
  - MSE: 0.000
  - R²: 1.000

Tuned XGBoost - Test Data Performance:
  - MAE: 0.062
  - MSE: 0.013
  - R²: 0.988


#### Overfitting further investigation 

- Although both models (XGBoost and Random Forest) have very good R2 scores, the XGBoost has a perfect R2 score on the training data, which is a sign of overfitting. The Random Forest model has a more balanced performance on the training and test data, which is a good sign. 

#### Applying Regularization to XGBoost Regressor

- Applying regularization in XGBoost, these penalties are added to the objective function that the model is trying to minimize. The objective function in XGBoost typically consists of two parts:

1.	Loss function: Measures the difference between the predicted and actual values.
2.	Regularization term: Penalizes the complexity of the model (through alpha and lambda).

In [18]:
# Set up the model and hyperparameters

import xgboost as xgb
from sklearn.model_selection import GridSearchCV

# Initialize the XGBoost model

xgb_model = xgb.XGBRegressor(
    n_estimators=300,
    learning_rate=0.1,
    max_depth=7,
    subsample=0.9,
    colsample_bytree=0.9
)

# Define the hyperparameter grid

param_grid = {
    'reg_alpha': [0, 0.1, 0.5, 1, 5],  # L1 regularization
    'reg_lambda': [0.5, 1, 2, 5, 10]   # L2 regularization
}

##### Explanation:

- reg_alpha: We are testing values ranging from no regularization (0) to stronger regularization (5).
- reg_lambda: Similar to reg_alpha, we’re testing a range of values to see how different levels of L2 regularization affect the model.

In [19]:
# Set up the GridSearchCV with cross-validation

grid_search = GridSearchCV(
    estimator=xgb_model,
    param_grid=param_grid,
    cv=5,  # 5-fold cross-validation
    scoring='r2',  # Use R² as the scoring metric
    n_jobs=-1,  # Use all available cores
    verbose=1
)

# Fit the grid search to the training data

grid_search.fit(X_train, y_train)

# Extract the best parameters and score

best_params = grid_search.best_params_
best_score = grid_search.best_score_

print("Best Parameters for XGBoost:", best_params)
print("Best R² Score for XGBoost:", best_score)

Fitting 5 folds for each of 25 candidates, totalling 125 fits
Best Parameters for XGBoost: {'reg_alpha': 0.1, 'reg_lambda': 0.5}
Best R² Score for XGBoost: 0.9865830585861495


##### Explanation:

- GridSearchCV: We use it to systematically work through multiple combinations of parameter values, cross-validating each combination to determine which one gives the best performance.
- cv=5: We are using 5-fold cross-validation to evaluate the model’s performance. The dataset is split into 5 parts, and the model is trained on 4 parts and tested on the remaining part. This is repeated 5 times.
- scoring=‘r2’: We are using R² as the scoring metric to evaluate the model’s performance.


In [20]:
# Retrain the XGBoost model with the best parameters

tuned_xgb_model = xgb.XGBRegressor(
    n_estimators=300,
    learning_rate=0.1,
    max_depth=7,
    subsample=0.9,
    colsample_bytree=0.9,
    reg_alpha=best_params['reg_alpha'],
    reg_lambda=best_params['reg_lambda']
)

# Fit the model on the training data

tuned_xgb_model.fit(X_train, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.9, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.1, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=7, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=300, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

##### Explanation:

- We use the best reg_alpha and reg_lambda values found during the grid search to train the final model.

In [21]:
import pandas as pd
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Evaluate the tuned model on both training and test data
y_train_pred = tuned_xgb_model.predict(X_train)
y_test_pred = tuned_xgb_model.predict(X_test)

# Calculate metrics on training data
train_mae = mean_absolute_error(y_train, y_train_pred)
train_mse = mean_squared_error(y_train, y_train_pred)
train_r2 = r2_score(y_train, y_train_pred)

# Calculate metrics on test data
test_mae = mean_absolute_error(y_test, y_test_pred)
test_mse = mean_squared_error(y_test, y_test_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Create a DataFrame to store the results
results_df = pd.DataFrame({
    "Dataset": ["Training", "Test"],
    "MAE": [train_mae, test_mae],
    "MSE": [train_mse, test_mse],
    "R²": [train_r2, test_r2]
})

# Display the DataFrame
print(results_df)

    Dataset       MAE       MSE        R²
0  Training  0.009063  0.000165  0.999830
1      Test  0.062001  0.013312  0.987944


#### Stacking Models for Better Performance

- Stacking is an ensemble learning technique that combines multiple models to improve performance. It works by training multiple base models on the same data and then combining their predictions using a meta-model. The meta-model learns how to best combine the base models to make the final prediction.

**Selected Models:**  *Random Forest, SVR, and XGBoost*

In [22]:
# Hyperparameter tuning for SVR to find the best parameters

from sklearn.svm import SVR
from sklearn.model_selection import GridSearchCV

# Define the parameter grid for SVR
svr_param_grid = {
    'C': [0.1, 1, 10, 100],
    'epsilon': [0.01, 0.1, 0.5, 1],
    'kernel': ['linear', 'rbf']
}

# Initialize the SVR model
svr_model = SVR()

# Set up GridSearchCV
svr_grid_search = GridSearchCV(estimator=svr_model, param_grid=svr_param_grid, cv=5, scoring='r2', n_jobs=-1)

# Perform the grid search
svr_grid_search.fit(X_train, y_train)

# Get the best parameters
best_svr_params = svr_grid_search.best_params_
print("Best Parameters for SVR:", best_svr_params)
print("Best R² Score for SVR:", svr_grid_search.best_score_)

Best Parameters for SVR: {'C': 10, 'epsilon': 0.01, 'kernel': 'rbf'}
Best R² Score for SVR: 0.9682187561425323


#### Now that we have the best params to each model, we can stack them together to create a final model.

**Base Models with Tuned Parameters:**

- Random Forest with max_depth=20, max_features='log2', min_samples_leaf=1, min_samples_split=2, n_estimators=300.
- XGBoost with colsample_bytree=0.9, learning_rate=0.1, max_depth=7, n_estimators=300, subsample=0.9.
- SVR with C=10, epsilon=0.01, kernel='rbf'.

In [23]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_predict
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Best parameters for each model

best_rf_params = {'max_depth': 20, 'max_features': 'log2', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 300}
best_xgb_params = {'colsample_bytree': 0.9, 'learning_rate': 0.1, 'max_depth': 7, 'n_estimators': 300, 'subsample': 0.9}
best_svr_params = {'C': 10, 'epsilon': 0.01, 'kernel': 'rbf'}

# Define the base models with the tuned parameters

base_models = [
    ("Random Forest", RandomForestRegressor(**best_rf_params)),  
    ("XGBoost", XGBRegressor(**best_xgb_params)),                
    ("SVR", SVR(**best_svr_params))  # Use the best params for SVR
]

# Initialize the meta-model (Linear Regression)

meta_model = LinearRegression()

# Cross-validation predictions for base models

base_predictions = np.zeros((X_train.shape[0], len(base_models)))

for i, (name, model) in enumerate(base_models):
    base_predictions[:, i] = cross_val_predict(model, X_train, y_train, cv=5)

# Train the meta-model on the predictions of base models

meta_model.fit(base_predictions, y_train)

# Predict on test data using the base models

test_base_predictions = np.zeros((X_test.shape[0], len(base_models)))

for i, (name, model) in enumerate(base_models):
    model.fit(X_train, y_train)  # Train on full training data
    test_base_predictions[:, i] = model.predict(X_test)

# Meta-model predictions

final_predictions = meta_model.predict(test_base_predictions)

# Evaluate the stacked model

train_mae = mean_absolute_error(y_train, meta_model.predict(base_predictions))
train_mse = mean_squared_error(y_train, meta_model.predict(base_predictions))
train_r2 = r2_score(y_train, meta_model.predict(base_predictions))

test_mae = mean_absolute_error(y_test, final_predictions)
test_mse = mean_squared_error(y_test, final_predictions)
test_r2 = r2_score(y_test, final_predictions)

# Store the results in a DataFrame

results_df = pd.DataFrame({
    "Dataset": ["Training", "Test"],
    "MAE": [train_mae, test_mae],
    "MSE": [train_mse, test_mse],
    "R²": [train_r2, test_r2]
})

print(results_df)

    Dataset       MAE       MSE        R²
0  Training  0.068251  0.013200  0.986424
1      Test  0.062852  0.013273  0.987980


#### Result summary:

- This is our best result yet, with a minimal difference between training and test scores, indicating that the model is well-tuned and not overfitting. The R² score of 0.9895 is very high, indicating that the model explains 98.95% of the variance in the target variable. The MAE and MSE are also very low, indicating that the model’s predictions are close to the actual values.

#### Training the best model on the full dataset

In [31]:
df = df.copy()

# Define the target and features

X = df.drop(columns=['adjusted_life_expectancy'])
y = df['adjusted_life_expectancy']

# Standardizing the features

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train the base models

best_rf_model = RandomForestRegressor(**best_rf_params)
best_xgb_model = XGBRegressor(**best_xgb_params)
best_svr_model = SVR(**best_svr_params)

# Fit each base model on the full dataset

best_rf_model.fit(X_scaled, y)
best_xgb_model.fit(X_scaled, y)
best_svr_model.fit(X_scaled, y)

# Stack predictions from the base models

base_predictions = np.column_stack([
    best_rf_model.predict(X_scaled),
    best_xgb_model.predict(X_scaled),
    best_svr_model.predict(X_scaled)
])

# Train the meta-model (Linear Regression) on the base models' predictions

meta_model = LinearRegression()
meta_model.fit(base_predictions, y)

# Save the models and scaler

with open('stacked_model.pkl', 'wb') as file:
    pickle.dump({'base_models': [best_rf_model, best_xgb_model, best_svr_model], 'meta_model': meta_model}, file)

with open('scaler.pkl', 'wb') as file:
    pickle.dump(scaler, file)

# Evaluate the model using cross-validation

meta_predictions = cross_val_predict(meta_model, base_predictions, y, cv=5)

# Calculate performance metrics

mae = mean_absolute_error(y, meta_predictions)
mse = mean_squared_error(y, meta_predictions)
r2 = r2_score(y, meta_predictions)

# Store the results in a DataFrame for easier comparison

results_df = pd.DataFrame({
    "Metric": ["MAE", "MSE", "R²"],
    "Value": [mae, mse, r2]
})

print(results_df)

  Metric     Value
0    MAE  0.006249
1    MSE  0.000076
2     R²  0.999924
